### PR TITLE
Scalarize some vector operations

### PR DESCRIPTION
**Pull request by nurmukhametov**
This PR implement scalarization of some vector operations that operate on vectors containing only the one actual value. It happens often in code generated to XE. With CPU, it is rare but happens either.

This depends on #2845

### CODE REVIEW

**Comment by aneshlya** [comment]
LLVM has `Scalarizer` pass https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Transforms/Scalar/Scalarizer.h. Is it something similar to what you're trying to implement? I checked it on one of your tests and the result is the same: https://ispc.godbolt.org/z/T8j4a6fPK

**Comment by nurmukhametov** [comment]
> LLVM has `Scalarizer` pass https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Transforms/Scalar/Scalarizer.h. Is it something similar to what you're trying to implement? I checked it on one of your tests and the result is the same: https://ispc.godbolt.org/z/T8j4a6fPK

It looks like this pass always scalarize vector operations. I am not sure that we want to do this. With this pass, around 30 lit-tests fail.

File: src/opt/ScalarizePass.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    OutZeroMask = SVI->getShuffleMaskForBitcode();
+
+    llvm::BinaryOperator::BinaryOps Opc = BO->getOpcode();
+    llvm::Value *Op1 = nullptr, *Op2 = nullptr;
+    if (match(BO, m_BinOp(m_InsertElt(m_Undef(), m_Value(Op1), m_ZeroInt()), m_Value(Vec))) ||
+        match(BO, m_BinOp(m_Value(Vec), m_InsertElt(m_Undef(), m_Value(Op2), m_ZeroInt())))) {
+
+        if (llvm::ConstantVector *CV = llvm::dyn_cast<llvm::ConstantVector>(Vec)) {
+            unsigned N = CV->getType()->getNumElements();
+            bool isTailPosion = true;
```
[the remark this answers is not in the document]
Comment (nurmukhametov) [reply src/opt/ScalarizePass.cpp]: Fixed

File: src/opt/ScalarizePass.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+#include "builtins-decl.h"
+
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Transforms/Utils/Local.h"
+
+namespace ispc {
+
+using namespace llvm::PatternMatch;
+
+static llvm::Value *lMatch(llvm::Instruction *inst) {
```
[the remark this answers is not in the document]
Comment (nurmukhametov) [reply src/opt/ScalarizePass.cpp]: Fixed

File: tests/lit-tests/scalarize.ll [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+; CHECK-NEXT:  %op2 = insertelement <4 x i32> undef, i32 %b, i32 0
+define <4 x i32> @nonconst(i32 %a, i32 %b) {
+  %op1 = insertelement <4 x i32> undef, i32 %a, i32 0
+  %op2 = insertelement <4 x i32> undef, i32 %b, i32 0
+  %sum = add <4 x i32> %op1, %op2
+  %shuffled = shufflevector <4 x i32> %sum, <4 x i32> poison, <4 x i32> zeroinitializer
+  ret <4 x i32> %shuffled
+}
+
+; not supported for values accross phi-nodes
```
[the remark this answers is not in the document]
Comment (nurmukhametov) [reply tests/lit-tests/scalarize.ll]: Fixed

File: src/opt/ScalarizePass.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#pragma once
+
+#include "ISPCPass.h"
+
+namespace ispc {
+
+// ScalarizePass change some vector operations to scalar ones.
```
[the remark this answers is not in the document]
Comment (nurmukhametov) [reply src/opt/ScalarizePass.h:13]: Done. You can also check the issue https://github.com/ispc/ispc/issues/2849

**Review by aneshlya — APPROVED** [review]
Looks good!

**Pull request merged**